### ISSUE TITLE
Sockets didn't get closed in SSLServer::process_and_close_socket().

### ISSUE THREAD

**Issue by axieyangb** [yhirose/cpp-httplib#706]
Hey @yhirose, I found the abnormal memory increase under the stress tests. The problem line is here: https://github.com/yhirose/cpp-httplib/blob/master/httplib.h#L5857
```
if(ssl){
// Missing the statements to close the socket.
}

```
The sockets keep open but never get closed. 
Comparing to the https://github.com/yhirose/cpp-httplib/blob/master/httplib.h#L4625-L4626. We should add 
`detail::shutdown_socket(sock);
  detail::close_socket(sock);` in line of 5857.

**Comment by yhirose** [comment]
@axieyangb, thanks for find the serious problem!
Could you try the stress tests again with the latest master on your machine?

1.

**Comment by axieyangb** [comment]
Sure, I am very glad to help.

**Comment by axieyangb** [comment]
No significant memory increase this time. But the memory still increases. 1000 QPS runs for 10 mins, the memory increases from 6 Mi to 18Mi.  Without fix before, the memory goes above 150 Mi.

**Comment by yhirose** [comment]
@axieyangb, thanks for the report. I'll reopen this issue. Please let me know whatever suspicious code. I'll also try to find where the leak is happening. Thanks!

**Comment by CaelumLaron** [comment]
How do you do your tests @axieyangb ? I'm with a problem and wanna check if it's the same

**Comment by axieyangb** [comment]
In my case, I used it in k8s. So I wrote a simple test to send multiple CURL commands in parallel to the pod (Comment out all the internal processing logic and only left the HTTPS server running) and observed the allocation of the pod resource.

**Comment by axieyangb** [comment]
Hey @yhirose , another suspicious code is here [not pretty sure]: `detail::ssl_delete(ctx_mutex_, ssl, ret);` https://github.com/yhirose/cpp-httplib/blob/master/httplib.h#L5854. If the ret is `false`, the ssl_delete() will skip calling the `SSL_shutdown(ssl);` https://github.com/yhirose/cpp-httplib/blob/master/httplib.h#L5623. And I quickly went through the function of `process_request()` and  tried to understand that in which situation, the `ret` will return false. I found several places return `false` but I am not sure that they should all skip the `SSL_shutdown(ssl);` in `ssl_delete()`. Maybe you can confirm that is right.

**Comment by axieyangb** [comment]
For example: https://github.com/yhirose/cpp-httplib/blob/master/httplib.h#L4134-L4138. If the server fails to write the boby. The `process_request()` returns false and the `SSL_shutdown(ssl)` will be skipped.

**Comment by yhirose** [comment]
@axieyangb, thanks for pointing it out. Could you check how often `detail::process_server_socket_ssl` returns `false` on your test environment? If it happens pretty often and increases memory usage, that could be it. But if it happens rarely and still increases memory usage, there should be another place where memory or resource leak is taking place. Thanks for your great help!

**Comment by yhirose** [comment]
@axieyangb, I did some research about it with Apache Bench command, and I am now able to reproduce it.

The interesting thing is that even if I changed [this line](https://github.com/yhirose/cpp-httplib/blob/3b29cd0bdce6f88d5c2cc9deb4670a629a76c4fe/httplib.h#L5881) to `detail::ssl_delete(ctx_mutex_, ssl, false);`, it doesn't make any difference. The memory usage still goes up, but the increase eventually stops and the final usage depends on how many **concurrent** requests being sent to the server.

I made the `example/hello.cc` sample server to use SSL and tested it with `top` command.
Here are my results:

### detail::ssl_delete(ctx_mutex_, ssl, ret)
If I run `ab -n 10000 -c 1 https://localhost:8080/hi`, the memory usage goes up to **1380K**.
If I run `ab -n 10000 -c 2 https://localhost:8080/hi`, the memory usage goes up to **1684K**. 
If I run `ab -n 10000 -c 3 https://localhost:8080/hi`, the memory usage goes up to **1980K**. 
If I run `ab -n 10000 -c 4 https://localhost:8080/hi`, the memory usage goes up to **2268K**. 
If I run `ab -n 10000 -c 5 https://localhost:8080/hi`, the memory usage goes up to **2548K**. 

### detail::ssl_delete(ctx_mutex_, ssl, false)
If I run `ab -n 10000 -c 1 https://localhost:8080/hi`, the memory usage goes up to **1392K**.
If I run `ab -n 10000 -c 2 https://localhost:8080/hi`, the memory usage goes up to **1684K**. 
If I run `ab -n 10000 -c 3 https://localhost:8080/hi`, the memory usage goes up to **1976K**. 
If I run `ab -n 10000 -c 4 https://localhost:8080/hi`, the memory usage goes up to **2272K**. 
If I run `ab -n 10000 -c 5 https://localhost:8080/hi`, the memory usage goes up to **2552K**. 

If I test with a non SSL server, memory usage stays around 1000K regardless how many concurrent requests are sent.

So I am suspicious that the memory leak is caused  by OpenSSL, because it happens even if we call `SSL_shutdown` and `SSL_free` in `detail::ssl_delete` properly. I don't think we can do anything for it. Good thing is that the memory usage is not huge anyway. 

Do you have any thoughts?

**Comment by yhirose** [comment]
@axieyangb, I'll close it because the first issue has been resolved with your help, and the 2nd issues is not actually memory leak. Please let me know if you find further information. Thanks!

**Comment by axieyangb** [comment]
Sorry for my late reply.   I used the golang  code to  simulate the concurrent calls to the HTTPS server. In my test case,  every second, the program will spawn 1000 threads  (-c 1000) to send the HTTP request. The memory increase is very obvious with that level of the load.

Another problem I found: The job list does not have the capacity limitation https://github.com/yhirose/cpp-httplib/blob/master/httplib.h#L555.   

In my e2e test, the client side will throw `too many openfile error` after running around 1 hour.

The reason is that the latter request comes into the server before the previous requests get finished, so more and more pending jobs are enqueued and the http sockets  are also remain opening.


 https://github.com/yhirose/cpp-httplib/blob/master/httplib.h#L4432

**Comment by axieyangb** [comment]
If you are OK, I can create another ticket about it.

**Comment by yhirose** [comment]
@axieyangb, thanks for the report. It's obviously not **memory leak**, but cpp-httplib cannot handle such concurrent requests quickly, end it ends up **having lots of pending requests** staying around... 

I humbly say that cpp-httplib is a **multi-threading, blocking, synchronous** server library, and this model is inadiquate to fix this kind of problems in nature... Apache cannot handle this c10k problem either due to the same reason. That's why people use NGINX or node.js. https://en.wikipedia.org/wiki/C10k_problem https://github.com/yhirose/cpp-httplib/issues/268

So if you need a HTTP server library which expects large amount of simultaneous requests, I recommend you to take a look at **non-blocking asynchronous** HTTP libraries such as [oat++](https://oatpp.io/) and [drogon](https://github.com/an-tao/drogon) instead.

By the way, the reason why I intentionally made such a **blocking, multi-threading, synchronous** HTTP library is that I needed something that can be easily embedded into our desktop products at work and easy to debug HTTP messages passing around. But there wasn't such a handy embeddable HTTP library at that time. So I made this single-header library for me. This means that it wasn't made for such a large scale HTTP web server that expects tons of simultaneous requests from outside world.

Hope it helps you to understand the purpose of this library!
